readium/navigator/src/main/java/org/readium/r2/navigator/media2/MediaSessionNavigator.kt

```diff
@@ -0,0 +1,272 @@
+package org.readium.r2.navigator.media2
+
+import android.app.PendingIntent
+import android.content.Context
+import androidx.media2.common.MediaMetadata
+import androidx.media2.common.SessionPlayer
+import androidx.media2.session.MediaSession
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import org.readium.r2.navigator.ExperimentalAudiobook
+import org.readium.r2.navigator.extensions.sum
+import org.readium.r2.navigator.extensions.time
+import org.readium.r2.shared.publication.*
+import org.readium.r2.shared.util.Try
+import org.readium.r2.shared.util.flatMap
+import timber.log.Timber
+import java.util.concurrent.Executors
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.ExperimentalTime
+
+/**
+ * An audiobook navigator to connect to a MediaSession from Jetpack Media2.
+ *
+ * Use [create] to get an instance for a given publication, and build a session from it
+ * with the [session] method. Apps are responsible for attaching this session to a service able to
+ * expose it.
+ *
+ * You can build a [MediaSessionNavigator] upon any Media2 [SessionPlayer] implementation
+ * providing [create] with it. If you don't, ExoPlayer will be used, without cache.
+ * Use [ExoPlayerFactory] to build a [SessionPlayer] based on ExoPlayer with caching capabilities.
+ */
+@ExperimentalAudiobook
+@OptIn(ExperimentalTime::class)
+class MediaSessionNavigator private constructor(
+    override val publication: Publication,
+    private val playerFacade: SessionPlayerFacade,
+    private val playerCallback: SessionPlayerCallback,
+    private val configuration: Configuration
+) : MediaNavigator {
+
+    private val coroutineScope: CoroutineScope = MainScope()
+
+    private fun <T> Flow<T>.stateInFirst(coroutineScope: CoroutineScope): StateFlow<T> =
+        stateIn(coroutineScope, SharingStarted.Lazily, runBlocking { first() })
+
+    private val totalDuration: Duration? =
+        this.playerFacade.playlist!!.durations?.sum()
```

Comment:
Are we sure `playlist` and others here in this file coming from `playerFacade` won't ever be null?

Reply:
Yes, I am.  `MediaSessionNavigator` is created only once the `SessionPlayer` behind is ready.